AttributeError: 'NoneType' object has no attribute 'items' in swinunetr algo

```
2024-04-30 01:07:45,323 - INFO - Auto3Dseg picked the following networks to ensemble:
2024-04-30 01:07:45,323 - INFO - swinunetr_0
2024-04-30 01:07:45,323 - INFO - Auto3Dseg ensemble prediction outputs will be saved in /opt/toolkit/tutorials/monai/auto3dseg/notebooks/my_workspace/ensemble_output.

Ensembling (rank 0)...:   0%|          | 0/8 [00:00<?, ?it/s]
monai.networks.nets.swin_unetr SwinUNETR.__init__:img_size: Argument `img_size` has been deprecated since version 1.3. It will be removed in version 1.5. The img_size argument is not required anymore and checks on the input size are run during forward().
monai.networks.nets.swin_unetr SwinUNETR.__init__:img_size: Argument `img_size` has been deprecated since version 1.3. It will be removed in version 1.5. The img_size argument is not required anymore and checks on the input size are run during forward().
monai.networks.nets.swin_unetr SwinUNETR.__init__:img_size: Argument `img_size` has been deprecated since version 1.3. It will be removed in version 1.5. The img_size argument is not required anymore and checks on the input size are run during forward().
monai.networks.nets.swin_unetr SwinUNETR.__init__:img_size: Argument `img_size` has been deprecated since version 1.3. It will be removed in version 1.5. The img_size argument is not required anymore and checks on the input size are run during forward().

Ensembling (rank 0)...:  12%|█▎        | 1/8 [00:02<00:14,  2.05s/it][rank1]: Traceback (most recent call last):
[rank1]:   File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
[rank1]:     return _run_code(code, main_globals, None,
[rank1]:   File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
[rank1]:     exec(code, run_globals)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/__main__.py", line 24, in <module>
[rank1]:     fire.Fire(
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 143, in Fire
[rank1]:     component_trace = _Fire(component, args, parsed_flag_args, context, name)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 477, in _Fire
[rank1]:     component, remaining_args = _CallAndUpdateTrace(
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 693, in _CallAndUpdateTrace
[rank1]:     component = fn(*varargs, **kwargs)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/ensemble_builder.py", line 603, in ensemble
[rank1]:     self.ensembler(pred_param=self.kwargs)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/ensemble_builder.py", line 214, in __call__
[rank1]:     pred = infer_instance.predict(predict_files=[file], predict_params=_param)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/bundle_gen.py", line 363, in predict
[rank1]:     inferer = self.get_inferer(**params)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/bundle_gen.py", line 351, in get_inferer
[rank1]:     return infer_class.InferClass(configs_path, *args, **kwargs)
[rank1]:   File "/opt/toolkit/tutorials/monai/auto3dseg/notebooks/my_workspace/swinunetr_0/scripts/infer.py", line 45, in __init__
[rank1]:     parser.read_config(config_file_)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/bundle/config_parser.py", line 319, in read_config
[rank1]:     content.update(self.load_config_files(f, **kwargs))
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/bundle/config_parser.py", line 426, in load_config_files
[rank1]:     for k, v in (cls.load_config_file(i, **kwargs)).items():
[rank1]: AttributeError: 'NoneType' object has no attribute 'items'
Ensembling (rank 0)...:  12%|█▎        | 1/8 [00:02<00:14,  2.05s/it][rank1]: Traceback (most recent call last):
[rank1]:   File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
[rank1]:     return _run_code(code, main_globals, None,
[rank1]:   File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
[rank1]:     exec(code, run_globals)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/__main__.py", line 24, in <module>
[rank1]:     fire.Fire(
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 143, in Fire
[rank1]:     component_trace = _Fire(component, args, parsed_flag_args, context, name)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 477, in _Fire
[rank1]:     component, remaining_args = _CallAndUpdateTrace(
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/fire/core.py", line 693, in _CallAndUpdateTrace
[rank1]:     component = fn(*varargs, **kwargs)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/ensemble_builder.py", line 603, in ensemble
[rank1]:     self.ensembler(pred_param=self.kwargs)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/ensemble_builder.py", line 214, in __call__
[rank1]:     pred = infer_instance.predict(predict_files=[file], predict_params=_param)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/bundle_gen.py", line 363, in predict
[rank1]:     inferer = self.get_inferer(**params)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/apps/auto3dseg/bundle_gen.py", line 351, in get_inferer
[rank1]:     return infer_class.InferClass(configs_path, *args, **kwargs)
[rank1]:   File "/opt/toolkit/tutorials/monai/auto3dseg/notebooks/my_workspace/swinunetr_0/scripts/infer.py", line 45, in __init__
[rank1]:     parser.read_config(config_file_)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/bundle/config_parser.py", line 319, in read_config
[rank1]:     content.update(self.load_config_files(f, **kwargs))
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/monai/bundle/config_parser.py", line 426, in load_config_files
[rank1]:     for k, v in (cls.load_config_file(i, **kwargs)).items():
[rank1]: AttributeError: 'NoneType' object has no attribute 'items'

Ensembling (rank 0)...:  25%|██▌       | 2/8 [00:02<00:07,  1.20s/it]
Ensembling (rank 0)...:  38%|███▊      | 3/8 [00:03<00:04,  1.09it/s]
Ensembling (rank 0)...:  50%|█████     | 4/8 [00:03<00:03,  1.27it/s]
[2024-04-30 01:07:49,575] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 4901 closing signal SIGTERM
[2024-04-30 01:07:49,575] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 4901 closing signal SIGTERM

[2024-04-30 01:07:49,790] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 1 (pid: 4902) of binary: /usr/bin/python
Traceback (most recent call last):
  File "/usr/local/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 347, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 834, in main
    run(args)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 825, in run
    elastic_launch(
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 137, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 271, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
monai.apps.auto3dseg FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-04-30_01:07:49
  host      : swqa-P1001CR-02
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 4902)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
[2024-04-30 01:07:49,790] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 1 (pid: 4902) of binary: /usr/bin/python
Traceback (most recent call last):
  File "/usr/local/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 347, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 834, in main
    run(args)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 825, in run
    elastic_launch(
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 137, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 271, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
monai.apps.auto3dseg FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-04-30_01:07:49
  host      : swqa-P1001CR-02
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 4902)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

--- a/auto3dseg/algorithm_templates/swinunetr/scripts/infer.py
+++ b/auto3dseg/algorithm_templates/swinunetr/scripts/infer.py
@@ -12,6 +12,7 @@
 import logging
 import os
 import sys
+from pathlib import Path
 from typing import Optional, Sequence, Union
 
 import torch
@@ -25,6 +26,7 @@ from monai.bundle import ConfigParser
 from monai.bundle.scripts import _pop_args, _update_args
 from monai.data import ThreadDataLoader, decollate_batch, list_data_collate
 from monai.inferers import sliding_window_inference
+from monai.utils.misc import ensure_tuple
 
 sys.path.insert(0, os.path.abspath(os.path.dirname(__file__)))
 if __package__ in (None, ""):
@@ -35,14 +37,19 @@ else:
 
 class InferClass:
     def __init__(self, config_file: Optional[Union[str, Sequence[str]]] = None, **override):
-        pre_operation(config_file, **override)
         logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
         _args = _update_args(config_file=config_file, **override)
         config_file_ = _pop_args(_args, "config_file")[0]
-
+        config_file_ = [
+            path
+            for path in ensure_tuple(config_file_)
+            if not (path.endswith("hyper_parameters.yaml") or Path(path).name.startswith(".") or path.endswith(".lock"))
+        ]
         parser = ConfigParser()
         parser.read_config(config_file_)
+        parser_hyper = pre_operation(config_file, **override)
+        parser.update(pairs=parser_hyper.config)
         parser.update(pairs=_args)
 
         data_file_base_dir = parser.get_parsed_content("data_file_base_dir")

--- a/auto3dseg/algorithm_templates/swinunetr/scripts/train.py
+++ b/auto3dseg/algorithm_templates/swinunetr/scripts/train.py
@@ -20,6 +20,7 @@ import random
 import sys
 import time
 import warnings
+from pathlib import Path
 from datetime import datetime
 from typing import Optional, Sequence, Union
 
@@ -28,6 +29,7 @@ import mlflow.pytorch
 import numpy as np
 import torch
 import torch.distributed as dist
+from filelock import FileLock
 import yaml
 from torch.nn.parallel import DistributedDataParallel
 from torch.nn.utils import clip_grad_norm_
@@ -46,6 +48,7 @@ from monai.data import DataLoader, partition_dataset
 from monai.inferers import sliding_window_inference
 from monai.metrics import compute_dice
 from monai.utils import RankFilter, set_determinism
+from monai.utils.misc import ensure_tuple
 
 if __package__ in (None, ""):
     from algo import auto_scale
@@ -105,26 +108,31 @@ class EarlyStopping:
 
 def pre_operation(config_file, **override):
     """Update the hyper_parameters.yaml based on GPU memory"""
-    rank = int(os.getenv("RANK", "0"))
-    if rank == 0:
-        if isinstance(config_file, str) and "," in config_file:
-            config_file = config_file.split(",")
+    if isinstance(config_file, str) and "," in config_file:
+        config_file = config_file.split(",")
 
-        for _file in config_file:
-            if "hyper_parameters.yaml" in _file:
+    for _file in config_file:
+        if _file.endswith("hyper_parameters.yaml"):
+            lock = FileLock(f"{_file}.lock")
+            with lock:
                 parser = ConfigParser(globals=False)
                 parser.read_config(_file)
-                auto_scale_allowed = override.get("auto_scale_allowed", parser["auto_scale_allowed"])
-                max_epoch = override.get("auto_scale_max_epochs", parser["auto_scale_max_epochs"])
-                if auto_scale_allowed:
-                    output_classes = parser["output_classes"]
-                    n_cases = parser["n_cases"]
-                    scaled = auto_scale(output_classes, n_cases, max_epoch)
-                    parser.update({"num_patches_per_iter": scaled["num_patches_per_iter"]})
-                    parser.update({"num_crops_per_image": scaled["num_crops_per_image"]})
-                    parser.update({"num_epochs": scaled["num_epochs"]})
-                    ConfigParser.export_config_file(parser.get(), _file, fmt="yaml", default_flow_style=None)
-    return
+            auto_scale_allowed = override.get("auto_scale_allowed", parser["auto_scale_allowed"])
+            max_epoch = override.get("auto_scale_max_epochs", parser["auto_scale_max_epochs"])
+            if auto_scale_allowed:
+                output_classes = parser["output_classes"]
+                n_cases = parser["n_cases"]
+                scaled = auto_scale(output_classes, n_cases, max_epoch)
+                parser.update({"num_patches_per_iter": scaled["num_patches_per_iter"]})
+                parser.update({"num_crops_per_image": scaled["num_crops_per_image"]})
+                parser.update({"num_epochs": scaled["num_epochs"]})
+                rank = int(os.getenv("RANK", "0"))
+                if rank == 0:
+                    with lock:
+                        ConfigParser.export_config_file(parser.get(), _file, fmt="yaml", default_flow_style=None)
+                if dist.is_initialized():
+                    dist.barrier()
+    return parser
 
 
 def run(config_file: Optional[Union[str, Sequence[str]]] = None, **override):
@@ -132,10 +140,7 @@ def run(config_file: Optional[Union[str, Sequence[str]]] = None, **override):
     if torch.cuda.device_count() > 1:
         dist.init_process_group(backend="nccl", init_method="env://")
         world_size = dist.get_world_size()
-        pre_operation(config_file, **override)
-        dist.barrier()
     else:
-        pre_operation(config_file, **override)
         world_size = 1
 
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
@@ -145,9 +150,16 @@ def run(config_file: Optional[Union[str, Sequence[str]]] = None, **override):
 
     _args = _update_args(config_file=config_file, **override)
     config_file_ = _pop_args(_args, "config_file")[0]
+    config_file_ = [
+        path
+        for path in ensure_tuple(config_file_)
+        if not (path.endswith("hyper_parameters.yaml") or Path(path).name.startswith(".") or path.endswith(".lock"))
+        ]
 
     parser = ConfigParser()
     parser.read_config(config_file_)
+    parser_hyper = pre_operation(config_file, **override)
+    parser.update(pairs=parser_hyper.config)
     parser.update(pairs=_args)
 
     amp = parser.get_parsed_content("amp")

--- a/auto3dseg/algorithm_templates/swinunetr/scripts/validate.py
+++ b/auto3dseg/algorithm_templates/swinunetr/scripts/validate.py
@@ -13,9 +13,9 @@ import csv
 import logging
 import os
 import sys
+from pathlib import Path
 from typing import Optional, Sequence, Union
 
-import numpy as np
 import torch
 import yaml
 
@@ -28,6 +28,7 @@ from monai.bundle.scripts import _pop_args, _update_args
 from monai.data import ThreadDataLoader, decollate_batch
 from monai.inferers import sliding_window_inference
 from monai.metrics import compute_dice
+from monai.utils.misc import ensure_tuple
 
 sys.path.insert(0, os.path.abspath(os.path.dirname(__file__)))
 if __package__ in (None, ""):
@@ -37,14 +38,19 @@ else:
 
 
 def run(config_file: Optional[Union[str, Sequence[str]]] = None, **override):
-    pre_operation(config_file, **override)
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
     _args = _update_args(config_file=config_file, **override)
     config_file_ = _pop_args(_args, "config_file")[0]
-
+    config_file_ = [
+        path
+        for path in ensure_tuple(config_file_)
+        if not (path.endswith("hyper_parameters.yaml") or Path(path).name.startswith(".") or path.endswith(".lock"))
+    ]
     parser = ConfigParser()
     parser.read_config(config_file_)
+    parser_hyper = pre_operation(config_file, **override)
+    parser.update(pairs=parser_hyper.config)
     parser.update(pairs=_args)
 
     data_file_base_dir = parser.get_parsed_content("data_file_base_dir")

